src/Workspaces/Remote/Core/Services/Options/FileWatcher.cs

```diff
@@ -0,0 +1,114 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Composition;
+using System.IO;
+using System.Threading;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.VisualStudio.CodingConventions;
+
+namespace Microsoft.CodeAnalysis.Editor.Options
+{
+    [Export(typeof(IFileWatcher)), Shared]
+    internal sealed class FileWatcher : IFileWatcher
+    {
+        private ImmutableDictionary<(string fileName, string directoryPath), FileSystemWatcher> _watchers = ImmutableDictionary<(string fileName, string directoryPath), FileSystemWatcher>.Empty;
+
+        [ImportingConstructor]
```

Comment:
you also said ResettableDelay is bad since it can introduce timing issue. this will certainly cause timing issue. and having 2 places for source of origin can cause weird race where one side said option is A and the other side said option is B.

Reply:
> you also said ResettableDelay is bad since it can introduce timing issue.

I was talking about a totally different timing issue.

> this will certainly cause timing issue. and having 2 places for source of origin can cause weird race where one side said option is A and the other side said option is B.

Yes, with this change it's possible for OOP to see different options on disk than Roslyn in-process. However, it's a workaround for the current situation where OOP *always* sees different options. This watcher is an implementation detail of Microsoft.VisualStudio.CodingConventions which will be removed as soon as we switch to the new compiler layer support.

Reply:
and why this timing issue is okay and ResettableDelay is bad when both will be removed when we switch to the new compiler layer support?